### PR TITLE
Advanced hash table

### PR DESCRIPTION
**Pull request by pllearns**
- WIP Failing test on hashFunction, will need to double check tests and methods

### CODE REVIEW

File: spec/hashTable_test.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  'use strict'
+
+  it('exists', () => {
+    expect(HashTable).to.be.a('function')
+  })
+
+  describe('put()', () => {
+    it('adds a key-value pair to the hash table.', () => {
+      const hash = new HashTable()
+      console.log(HashTable);
```
Comment (Anasauce) [inline spec/hashTable_test.js:17]: I'm assuming this is a WIP PR...but just in case you missed this console.log :)

File: src/hashTable.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    this.key = key,
+    this.value = value,
+    this.next = null
+  }
+}
+
+export default class HashTable {
+  constructor() {
+    this._length = 0
+    this.valueStore = []
```
Comment (Anasauce) [inline src/hashTable.js]: if you are encapsulating ```_length``` here on line 11 you would probably want to do that on line 12 to ```valueStore``` too. I'm not too familiar with this naming convention...had to ask @jrob about it.

Would love to chat about you reasoning if there is time. Or once you get the tests to pass.

File: src/hashTable.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  hashFunction(key) {
+    let sum = 0
```
Comment (Anasauce) [inline src/hashTable.js]: a higher order function could be used here...like map()

**Comment by Anasauce** [comment]
11 files changed! ☹️

File: spec/hashTable_test.js (hunk 1 above)
Comment (pllearns) [reply spec/hashTable_test.js:17]: Thanks!

File: src/hashTable.js (hunk 2 above)
Comment (pllearns) [reply src/hashTable.js]: All but one test is passing, but we can chat in the afternoon if you like :)

File: src/hashTable.js (hunk 3 above)
Comment (pllearns) [reply src/hashTable.js]: true, but I wasn't sure if that was allowed since we are supposed to be breaking down the creation of these data structures. Thanks for suggesting that! 👍

File: spec/hashTable_test.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      hash.put('name', 'philliplorenzo')
+      hash.put('42.0', 'renzo')
+      expect(hash.contains('42.0')).to.be.true
+    })
+  })
+
+  describe('get()', () => {
+    it('returns the data associated with key.', () => {
+      const hash = new HashTable()
+      hash.put('name', 'batman')
```
Comment (jrobcodes) [inline spec/hashTable_test.js:27]: The test need not be this verbose - we can insert one thing, and retrieve one thing, and that should be enough to validate the functionality.  With all of these inserts, it makes it hard to see what we are actually validating.

File: spec/hashTable_test.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      console.log(HashTable);
+      hash.put('name', 'philliplorenzo')
+      hash.put('42.0', 'renzo')
+      expect(hash.contains('42.0')).to.be.true
+    })
+  })
+
+  describe('get()', () => {
+    it('returns the data associated with key.', () => {
+      const hash = new HashTable()
```
Comment (jrobcodes) [inline spec/hashTable_test.js:26]: Consider using whitespace between variable declaration, test setup, and expectations to improve readability.

Comment (jrobcodes) [reply spec/hashTable_test.js:26]: This comment applies to all tests in this PR

File: spec/hashTable_test.js (hunk 4 above)
Comment (jrobcodes) [reply spec/hashTable_test.js:27]: This comment applies to all tests in this PR

File: spec/hashTable_test.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    expect(hash.contains('name')).to.be.true
+    hash.remove('matrix')
+    hash.remove('name')
+    expect(hash.contains('matrix')).to.be.false
+    expect(hash.contains('name')).to.be.false
+  })
+})
+
+  describe('size()', () => {
+    it('returns the number of key-value pairs in the hash table.', () => {
```
Comment (jrobcodes) [inline spec/hashTable_test.js:71]: I would also test the case where the hash is empty.

File: spec/hashTable_test.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  })
+
+  describe('hashFunction()', () => {
+    it('generates a hash for the key "name".', () => {
+      const hash = new HashTable()
+      hash.put('name', 'john')
+      hash.put('42.0', 'wick')
+      hash.put('keanu', 'reeves')
+      hash.put('matrix', 'anderson')
+      expect(hash.hashFunction()).to.eql('oiushdfoi')
```
Comment (jrobcodes) [inline spec/hashTable_test.js:88]: I don't understand how this test is testing what it describes?  No key is being passed to `hashFunction`...

File: src/hashTable.js (hunk 2 above)
Comment (jrobcodes) [reply src/hashTable.js]: If you choose to use the `_` naming convention to indicate "private" data/instance variables, you should do so consistently (i.e. `valueStore` should be `_valueStore`).

File: src/hashTable.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    for (let i = 0; i < key.length; i++) {
+      sum += key.charCodeAt(i)
+    }
+    return sum % 3550
```
Comment (jrobcodes) [inline src/hashTable.js]: Why `3550`? The use of an intention revealing constant, instead of a "magic number", might help to make this code clearer.

File: src/hashTable.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+          if (currentNode.key === key) {
+            return currentNode.value
+          } else {
+            currentNode = currentNode.next
+          }
+        }
+        if (currentNode.key === key) {
+          return currentNode.value
+        } else {
+          return "no value here"
```
Comment (jrobcodes) [inline src/hashTable.js]: I would not recommend returning a string here, as this would not necessarily indicate failure to the caller.  What if strings are the values that are being held in this hash, and someone stored the string "no value here"? How would the client determine that this was a failure?

Consider returning `null` when a value is not found.

File: src/hashTable.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    ++this._length
+  }
+
+  size() {
+    return this._length
+  }
+
+  get(key) {
+    let hash = this.hashFunction(key)
+    if (this.valueStore[hash]) {
```
Comment (jrobcodes) [inline src/hashTable.js]: I would make this whole block more concise:
```js
let current = this.valueStore[ this.hashFunction( key ) ]

while( current !== null && current.key !== key ) {
  current = current.next
}

return current === null ? null : current.value
```

File: src/hashTable.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+          return currentNode.value
+        } else {
+          return "no value here"
+        }
+      }
+    }
+  }
+
+  contains(key){
+    let hash = this.hashFunction(key)
```
Comment (jrobcodes) [inline src/hashTable.js]: `contains` could be made more concise using the code suggested for `get`, above.

File: src/hashTable.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    } else {
+      return false
+    }
+  }
+
+  iterate(callback) {
+    let currentNode = this.valueStore
+    for (let i = 0; i < currentNode.length; i++) {
+      if (currentNode[i]) {
+        callback(currentNode[i].key, currentNode[i].value)
```
Comment (jrobcodes) [inline src/hashTable.js]: The two calls to `callback` are a little confusing in this block, as is the identifier `currentNode` - it's not a Node, it's a value store...  Since we already know that `currentNode[ i ]` is a node (not null), we could just do the following:
```js
for( let index = 0; index < this.valueStore.length; index++ ) {
  let current = this.valueStore[ index ]

  while( current !== null && current !== undefined ) {
    callback( current.key, current.value )
    current = current.next
  } 
}
```

File: src/hashTable.js [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  }
+
+  chain(hash, newNode) {
+    newNode.next = this.valueStore[hash]
+    this.valueStore[hash] = newNode
+  }
+
+  save(hash, key, value){
+    elements[hash] = [pairing(key, value)]
```
Comment (jrobcodes) [inline src/hashTable.js]: Where is `pairing` defined?  What is `elements`?  Where is `count` defined? Is `save` part of the spec for a hash table?
